src/NHibernate/Loader/Loader.cs

```diff
@@ -992,7 +992,7 @@ private void CheckVersion(int i, IEntityPersister persister, object id, object e
 			}
 			else
 			{
-				obj = session.Instantiate(instanceClass, key.Identifier);
+				obj = session.Instantiate(instanceClass.EntityName, key.Identifier);
```

Comment:
Indeed, this change defeats the purpose of your patch: avoiding to resolve again the persister. The naming could be improved for code clarity: `instanceClass` should be renamed `concretePersister` and `GetRowPersister` would be renamed `GetConcretePersister`.